engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoMultiType.java

```diff
@@ -0,0 +1,182 @@
+package org.enso.interpreter.runtime.data;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.GenerateUncached;
+import com.oracle.truffle.api.dsl.NeverDefault;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.nodes.Node;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.concurrent.ConcurrentHashMap;
+import org.enso.interpreter.runtime.EnsoContext;
+
+/**
+ * Internal representation of {@code Type[]} that supports identity comparision with {@code ==} to
+ * support inline caching of {@link EnsoMultiValue}. This is a separate and hidden concept from
+ * {@link Type} which is used thru out the Enso codebase.
+ *
+ * <p>Think twice before opening this type to public!
+ */
+final class EnsoMultiType {
+  private static final ConcurrentHashMap<EnsoMultiType, EnsoMultiType> ALL_TYPES =
+      new ConcurrentHashMap<>();
+
+  @CompilerDirectives.CompilationFinal(dimensions = 1)
+  private final Type[] types;
+
+  private EnsoMultiType(Type[] types) {
+    this.types = types;
+  }
+
+  @CompilerDirectives.TruffleBoundary
+  static EnsoMultiType findOrCreateSlow(Type[] types, int from, int to) {
+    var mt = new EnsoMultiType(Arrays.copyOfRange(types, from, to));
+    return ALL_TYPES.computeIfAbsent(mt, java.util.function.Function.identity());
+  }
+
+  final int typesLength() {
+    return types.length;
+  }
+
+  final Type firstType() {
+    return types[0];
+  }
+
+  private int find(EnsoContext ctx, Type type) {
+    for (var i = 0; i < types.length; i++) {
+      for (var t : types[i].allTypes(ctx)) {
+        if (t == type) {
+          return i;
+        }
+      }
+    }
+    return -1;
+  }
+
+  @Override
+  @CompilerDirectives.TruffleBoundary
+  public int hashCode() {
+    int hash = 7;
+    hash = 89 * hash + Arrays.deepHashCode(this.types);
+    return hash;
+  }
+
+  @Override
+  @CompilerDirectives.TruffleBoundary
+  public boolean equals(Object obj) {
+    if (this == obj) {
+      return true;
+    }
+    if (obj == null) {
+      return false;
+    }
+    if (getClass() != obj.getClass()) {
+      return false;
+    }
+    final EnsoMultiType other = (EnsoMultiType) obj;
+    return Arrays.deepEquals(this.types, other.types);
```

Comment:
Shouldn't only `deepEquals` be under `TruffleBoundary`?

Reply:
Yes/No.
- technically the previous statements are eligible for PE
- however I don't want PE to call into `equals` and `hashCode` at all
- only callbacks from `ALL_TYPES.computeIfAbset` should access these two methods
- only `findOrCreateSlow` (itself by a boundary) should be calling `ALL_TYPES.computeIfAbset`
- so maybe we don't need `@TruffleBoundary` here at all

However excluding code from PE multiple times hurts nothing...